File: test-suite-kotlin/src/test/kotlin/io/micronaut/security/audit/docs/createdby/UserAuditingTest.kt
package io.micronaut.security.audit.docs.createdby

import io.micronaut.context.annotation.Property
import io.micronaut.context.annotation.Replaces
import io.micronaut.context.annotation.Requires
import io.micronaut.data.jdbc.annotation.JdbcRepository
import io.micronaut.data.model.query.builder.sql.Dialect
import io.micronaut.data.repository.CrudRepository
import io.micronaut.security.authentication.Authentication
import io.micronaut.security.utils.DefaultSecurityService
import io.micronaut.security.utils.SecurityService
import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import jakarta.inject.Inject
import jakarta.inject.Singleton
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.util.*

@Property(name = "datasources.default.dialect", value = "H2")
@Property(name = "datasources.default.schema-generate", value = "CREATE_DROP")
@Property(name = "datasources.default.url", value = "jdbc:h2:mem:devDb;LOCK_TIMEOUT=10000;DB_CLOSE_ON_EXIT=FALSE")
@Property(name = "datasources.default.username", value = "sa")
@Property(name = "datasources.default.driver-class-name", value = "org.h2.Driver")
@Property(name = "spec.name", value = "UserAuditingTest")
@MicronautTest(transactional = false)
class UserAuditingTest {

    @Inject
    var bookRepository: BookRepository? = null

    @Test
    fun testCreatedByUpdatedByPopulatedOnSave() {
        var book = Book()
        book.title = "Tropic of Cancer"
        book.author = "Henry Miller"
        book = bookRepository!!.save(book)
        Assertions.assertNotNull(book.id)
        Assertions.assertEquals("sherlock", book.creator)
        Assertions.assertEquals("sherlock", book.editor)
    }

    @Requires(property = "spec.name", value = "UserAuditingTest")
    @Replaces(DefaultSecurityService::class)
    @Singleton
    internal class MockSecurityService : SecurityService {
        override fun username(): Optional<String> {
            return Optional.of("sherlock")
        }

        override fun getAuthentication(): Optional<Authentication> {
            return Optional.of(Authentication.build(username().orElseThrow()))
        }

        override fun isAuthenticated(): Boolean {
            return true
        }

        override fun hasRole(role: String): Boolean {
            return false
        }
    }

    @Requires(property = "spec.name", value = "UserAuditingTest")
    @JdbcRepository(dialect = Dialect.H2)
    interface BookRepository : CrudRepository<Book?, Long?>
}

File: test-suite-python/src/test/python/micronaut/security/audit/docs/createdby/UserAuditingTest.py
from typing import Annotated

from jakarta.inject import Inject, Singleton
from java.util import Optional
from micronaut.context.annotation import Property, Replaces, Requires
from micronaut.data.jdbc.annotation import JdbcRepository
from micronaut.data.repository import CrudRepository
from micronaut.security.authentication import Authentication
from micronaut.security.utils import DefaultSecurityService, SecurityService
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .Book import Book


@Property(name="datasources.default.dialect", value="H2")
@Property(name="datasources.default.schema-generate", value="CREATE_DROP")
@Property(name="datasources.default.url", value="jdbc:h2:mem:devDb;LOCK_TIMEOUT=10000;DB_CLOSE_ON_EXIT=FALSE")
@Property(name="datasources.default.username", value="sa")
@Property(name="datasources.default.driver-class-name", value="org.h2.Driver")
@Property(name="spec.name", value="UserAuditingTest")
@MicronautTest(transactional=False)
class UserAuditingTest:
    bookRepository: Annotated["BookRepository", Inject]

    @Test
    def test_created_by_updated_by_populated_on_save(self):
        book = Book(None, "Tropic of Cancer", "Henry Miller", None, None)
        book = self.bookRepository.save(book)
        assert book.id is not None
        assert book.creator == "sherlock"
        assert book.editor == "sherlock"


@Requires(property="spec.name", value="UserAuditingTest")
@Replaces(DefaultSecurityService)
@Singleton
class MockSecurityService(SecurityService):

    def username(self) -> Optional[str]:
        return Optional.of("sherlock")

    def getAuthentication(self) -> Optional[Authentication]:
        return Optional.of(Authentication.build(self.username().orElseThrow()))

    def isAuthenticated(self) -> bool:
        return True

    def hasRole(self, role: str) -> bool:
        return False


@Requires(property="spec.name", value="UserAuditingTest")
@JdbcRepository(dialect="H2")
class BookRepository(CrudRepository[Book, int]):
    ...
